File: Dai-Hentai/ViewControllers/GalleryViewController/GalleryCollectionViewHandler.m
//
//  GalleryCollectionViewHandler.m
//  Dai-Hentai
//
//  Created by DaidoujiChen on 2017/2/16.
//  Copyright © 2017年 DaidoujiChen. All rights reserved.
//

#import "GalleryCollectionViewHandler.h"

@implementation GalleryCollectionViewHandler

#pragma mark - UICollectionViewDataSource

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return [self.delegate totalCount];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    [self.delegate toggleLoadPages];
    GalleryCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"GalleryCell" forIndexPath:indexPath];
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView willDisplayCell:(GalleryCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath {
    [self.delegate toggleDisplayImageAt:indexPath inCell:cell];
}

#pragma mark - UICollectionViewDelegateFlowLayout

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return [self.delegate cellSizeAt:indexPath inCollectionView:collectionView];
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return 0;
}
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return 0;
}

#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UICollectionView *)collectionView {
    [self.delegate userCurrentIndex:[self userCurrentIndexPath:collectionView].row + 1];
}

#pragma mark - Private Instance Method

// 算出使用者正看到幾頁
- (NSIndexPath *)userCurrentIndexPath:(UICollectionView *)collectionView {
    CGRect visibleRect;
    visibleRect.origin = collectionView.contentOffset;
    visibleRect.size = collectionView.bounds.size;
    CGPoint visiblePoint = CGPointMake(CGRectGetMidX(visibleRect), CGRectGetMidY(visibleRect));
    NSIndexPath *visibleIndexPath = [collectionView indexPathForItemAtPoint:visiblePoint];
    return visibleIndexPath;
}

@end

File: Dai-Hentai/ViewControllers/GalleryViewController/GalleryCollectionViewHandler.swift
//
//  GalleryCollectionViewHandler.swift
//  Dai-Hentai
//
//  Created by DaidoujiChen on 2017/2/16.
//  Copyright © 2017年 DaidoujiChen. All rights reserved.
//

import Foundation


@objc protocol GalleryCollectionViewHandlerDelegate {
    func totalCount() -> Int
    func toggleLoadPages()
    func toggleDisplayImage(at indexPath: IndexPath?, in cell: GalleryCell?)
    func cellSize(at indexPath: IndexPath?, in collectionView: UICollectionView?) -> CGSize
    func userCurrentIndex(_ index: Int)
}


@objc class GalleryCollectionViewHandler: NSObject, UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
    @objc var delegate: GalleryCollectionViewHandlerDelegate?
    
// MARK: - UICollectionViewDataSource
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return delegate?.totalCount() ?? 0
    }

    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        delegate?.toggleLoadPages()
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "GalleryCell", for: indexPath) as? GalleryCell
        return cell!
    }

    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
        delegate?.toggleDisplayImage(at: indexPath, in: cell as? GalleryCell)
    }

// MARK: - UICollectionViewDelegateFlowLayout
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return delegate?.cellSize(at: indexPath, in: collectionView) ?? CGSize()
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return 0
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return 0
    }

// MARK: - UIScrollViewDelegate
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        guard let collectionView = scrollView as? UICollectionView else { return }
        delegate?.userCurrentIndex((userCurrentIndexPath(collectionView)?.row ?? 0) + 1)
    }

// MARK: - Private Instance Method
    
    // 算出使用者正看到幾頁
    func userCurrentIndexPath(_ collectionView: UICollectionView?) -> IndexPath? {
        var visibleRect = CGRect()
        visibleRect.origin = collectionView?.contentOffset ?? CGPoint.zero
        visibleRect.size = collectionView?.bounds.size ?? CGSize.zero
        let visiblePoint = CGPoint(x: visibleRect.midX, y: visibleRect.midY)
        let visibleIndexPath = collectionView?.indexPathForItem(at: visiblePoint)
        return visibleIndexPath
    }
}
